### PR TITLE
SDK-1901: Repeatable HTTP client

### PR DESCRIPTION
**Pull request by sergeyspryker**
https://spryker.atlassian.net/browse/SDK-1901

what's done:
- Added repeatable HTTP client that makes several attempts to send request in case of connection failures

### CODE REVIEW

**Comment by dereuromark** [comment]
HTTP is an acronym, that should be reflected in title, changelog etc.

File: src/Upgrade/Infrastructure/VersionControlSystem/SourceCodeProvider/GitHub/GitHubClientFactory.php [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+use Github\HttpClient\Builder;
+use Psr\Http\Client\ClientInterface;
+use Upgrade\Infrastructure\Configuration\ConfigurationProvider;
+
+class GitHubClientFactory
+{
+    /**
+     * @var \Github\Client|null
+     */
+    protected ?Client $client = null;
```
Comment (vol4onok) [inline src/Upgrade/Infrastructure/VersionControlSystem/SourceCodeProvider/GitHub/GitHubClientFactory.php:23]: protected ?Client $client;

File: src/Upgrade/Infrastructure/VersionControlSystem/SourceCodeProvider/GitLab/GitLabClientFactory.php [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+use Gitlab\HttpClient\Builder;
+use Psr\Http\Client\ClientInterface;
+use Upgrade\Infrastructure\Configuration\ConfigurationProvider;
+
+class GitLabClientFactory
+{
+    /**
+     * @var \Gitlab\Client|null
+     */
+    protected ?Client $client = null;
```
Comment (vol4onok) [inline src/Upgrade/Infrastructure/VersionControlSystem/SourceCodeProvider/GitLab/GitLabClientFactory.php:22]: protected ?Client $client

**Review by DmytroKlymanSpryker — APPROVED** [review]

File: src/Upgrade/Infrastructure/VersionControlSystem/SourceCodeProvider/GitHub/GitHubClientFactory.php [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+            $this->client = $this->createClient();
+        }
+
+        return $this->client;
+    }
+
+    /**
+     * @return \Github\Client
+     */
+    protected function createClient(): Client
```
Comment (vol4onok) [inline src/Upgrade/Infrastructure/VersionControlSystem/SourceCodeProvider/GitHub/GitHubClientFactory.php:60]: move this init to constructor

File: src/Upgrade/Infrastructure/HttpClient/RepeatableHttpClient.php [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+            $response = $this->doRequest($request);
+
+            if ($response !== null) {
+                return $response;
+            }
+
+            ++$currentAttempt;
+        }
+
+        throw new RemoteServerUnreachableException(sprintf('Server `%s` is unreachable.', $request->getUri()));
```
Comment (vol4onok) [inline src/Upgrade/Infrastructure/HttpClient/RepeatableHttpClient.php:70]: catch this exception and send message <error> with isSuccessful false

File: src/Upgrade/Infrastructure/VersionControlSystem/SourceCodeProvider/GitHub/GitHubClientFactory.php (hunk 3 above)
Comment (pavelmaksimov25) [reply src/Upgrade/Infrastructure/VersionControlSystem/SourceCodeProvider/GitHub/GitHubClientFactory.php:60]: That would be unti-pattern. Please, don't. Current logic is fine.

**Pull request merged**